pkg/vmconfig/virtualcontroller/virtualcontroller_reconciler.go

```diff
@@ -0,0 +1,873 @@
+			},
+		}
+	case vmopv1.SCSIControllerTypeBusLogic:
+		controller = &vimtypes.VirtualBusLogicController{
+			VirtualSCSIController: vimtypes.VirtualSCSIController{
+				VirtualController: vimtypes.VirtualController{
+					VirtualDevice: vimtypes.VirtualDevice{
+						ControllerKey: pciController.Key,
+					},
+					BusNumber: spec.BusNumber,
+				},
+				SharedBus: sharingModeToSCSIVimTypes(ctx, spec.SharingMode),
+			},
+		}
+	case vmopv1.SCSIControllerTypeLsiLogic:
+		controller = &vimtypes.VirtualLsiLogicController{
+			VirtualSCSIController: vimtypes.VirtualSCSIController{
+				VirtualController: vimtypes.VirtualController{
+					VirtualDevice: vimtypes.VirtualDevice{
+						ControllerKey: pciController.Key,
+					},
+					BusNumber: spec.BusNumber,
+				},
+				SharedBus: sharingModeToSCSIVimTypes(ctx, spec.SharingMode),
+			},
+		}
+	case vmopv1.SCSIControllerTypeLsiLogicSAS:
+		controller = &vimtypes.VirtualLsiLogicSASController{
+			VirtualSCSIController: vimtypes.VirtualSCSIController{
+				VirtualController: vimtypes.VirtualController{
+					VirtualDevice: vimtypes.VirtualDevice{
+						ControllerKey: pciController.Key,
+					},
+					BusNumber: spec.BusNumber,
+				},
+				SharedBus: sharingModeToSCSIVimTypes(ctx, spec.SharingMode),
+			},
+		}
+	}
+
+	if spec.PCISlotNumber != nil {
+		controller.GetVirtualSCSIController().GetVirtualDevice().SlotInfo =
+			&vimtypes.VirtualDevicePciBusSlotInfo{
+				PciSlotNumber: *spec.PCISlotNumber,
+			}
+	}
+
+	return controller
+}
+
+// sharingModeToSCSIVimTypes converts the vmopv1 sharing mode to the
+// vimtypes VirtualSCSISharing enum.
+func sharingModeToSCSIVimTypes(
+	ctx context.Context,
+	sharingMode vmopv1.VirtualControllerSharingMode,
+) vimtypes.VirtualSCSISharing {
+
+	switch sharingMode {
+	case vmopv1.VirtualControllerSharingModeNone:
+		return vimtypes.VirtualSCSISharingNoSharing
+	case vmopv1.VirtualControllerSharingModePhysical:
+		return vimtypes.VirtualSCSISharingPhysicalSharing
+	case vmopv1.VirtualControllerSharingModeVirtual:
+		return vimtypes.VirtualSCSISharingVirtualSharing
+	default:
+		pkglog.FromContextOrDefault(ctx).
+			Error(errInvalidSharingMode, "defaulting to noSharing", "sharingMode", sharingMode)
+		return vimtypes.VirtualSCSISharingNoSharing
+	}
+}
+
+// sharingModeToNVMEControllerSharing converts the vmopv1 sharing mode to the
+// vimtypes VirtualNVMEControllerSharing enum.
+func sharingModeToNVMEControllerSharing(
+	ctx context.Context,
+	sharingMode vmopv1.VirtualControllerSharingMode,
+) vimtypes.VirtualNVMEControllerSharing {
+
+	switch sharingMode {
+	case vmopv1.VirtualControllerSharingModeNone:
+		return vimtypes.VirtualNVMEControllerSharingNoSharing
+	case vmopv1.VirtualControllerSharingModePhysical:
+		return vimtypes.VirtualNVMEControllerSharingPhysicalSharing
+	default:
+		pkglog.FromContextOrDefault(ctx).
+			Error(errInvalidSharingMode, "defaulting to noSharing", "sharingMode", sharingMode)
```

Comment:
if there is no switch case for VirtualControllerSharingModeVirtual. 

If the user sets the value to VirtualControllerSharingModeVirtual and our validating webhook does not catch that. We are essentially changing it to VirtualNVMEControllerSharingNoSharing here. I am wondering if this could cause intended behavior.